### PR TITLE
[WIP] Store version history for SiteConfig 

### PR DESCRIPTION
**Pull request by mikenz**
And allow viewing diffs between versions. 

Partial solution to https://github.com/silverstripe/silverstripe-siteconfig/issues/1

Draft/Published/Workflow for siteconfig is not required for our implementation so I won't be contributing a PR for that in the foreseeable future.

### CODE REVIEW

**Comment by tractorcow** [comment]
Wow this is amazing. I need to test this out. ;)

**Comment by tractorcow** [comment]
@mikenz had a chat with @chillu and we like this feature but it may be a bit late / low priority to go into 4.0.0, so would 4.1 be a safe enough milestone for this? Apologies if you are forced to use a fork for the time being. (hopefully we can fork 4.0 from 4 so we can merge it sooner rather than later).

**Comment by mikenz** [comment]
No stress here - we've been on our own fork of SiteConfig for ages :-) Looks like I'll be adding a Publish action after all, so I'll just keep adding code here.

**Comment by tractorcow** [comment]
A quick question for @silverstripe/core-team ; Are we happy to version SiteConfig? Does it make sense to everyone that this is the default? If so I will go ahead with preparing this for merge. I'll likely branch and do my own cleanup / testing. :)

**Comment by robbieaverill** [comment]
I like the idea. You might want to change site configuration settings as part of a campaign, which you couldn't do without versioning.

File: _config/config.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+Name: siteconfig
+After:
+ - 'versionied/*'
```
Comment (chillu) [inline _config/config.yml]: typo

**Comment by chillu** [comment]
I'm OK with versioning `SiteConfig`, makes sense given our recommendations elsewhere. You could argue that it's an API change (change to `SiteConfig_Live` table name in SQL queries), but I don't consider table names part of our direct API surface - you should be using the ORM which would transparently handle that.

But the specific implementation requires a lot of custom controller PHP. Which is the fault of our view layer approach rather than anything Mike could've done better. @robbieaverill will work on a history viewer for content blocks in January, which is intended to be a general purpose viewer that should be useful for SiteConfig: https://github.com/silverstripe/silverstripe-versioned/issues/37. Maybe we hold off until then?

**Comment by dhensby** [comment]
I'd like to get `SiteConfig` versioned - it feels more inline with our "version everything" approach to 4.x

If we're going to be creating a general purpose history viewer, maybe that part of this PR could be stripped out for now and we implement the general purpose viewer on SiteConfig when it's ready?

**Comment by mikenz** [comment]
I'll also have a look at what other changes I have in our live branch that I haven't pushed to this one. From memory there have been a bunch of small changes.

**Comment by tractorcow** [comment]
Putting this in blocked for now; we are in the process of developing our core "version viewer", which we will want to apply versioned siteconfig to once it's ready.

**Comment by chillu** [comment]
Since the version viewer won't be available until Feb 2018 in core, I'll remove this from the milestone.

**Comment by maxime-rainville** [comment]
> Since the version viewer won't be available until Feb 2018 in core, I'll remove this from the milestone.

@chillu the eternal optimist ;-)

Everyone seem to be quite stoked about this PR. Sounds like a missed opportunity if we let it go stale. Shall we move it to the backlog with the aim to ship it in SS4.4?

**Comment by maxime-rainville** [comment]
I'll close this PR for lack of activity. I'm happy to reopen if someone think it can be salvaged.

**Pull request closed without merging**